atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java

```diff
@@ -245,29 +250,38 @@ private CassandraClientPoolImpl(
     private void tryInitialize() {
         cassandra.cacheInitialCassandraHosts();
 
-        refreshPoolFuture = refreshDaemon.scheduleWithFixedDelay(
-                () -> {
-                    try {
-                        refreshPool();
-                    } catch (Throwable t) {
-                        log.warn(
-                                "Failed to refresh Cassandra KVS pool."
-                                        + " Extended periods of being unable to refresh will cause perf degradation.",
-                                t);
-                    }
-                },
-                config.poolRefreshIntervalSeconds(),
-                config.poolRefreshIntervalSeconds(),
-                TimeUnit.SECONDS);
-
         // for testability, mock/spy are bad at mockability of things called in constructors
         if (startupChecks == StartupChecks.RUN) {
             runOneTimeStartupChecks();
         }
-        refreshPool(); // ensure we've initialized before returning
+        runAndScheduleNextRefresh(0); // ensure we've initialized before returning
         metrics.registerAggregateMetrics(blacklist::size);
     }
 
+    private void runAndScheduleNextRefresh(int attempt) {
+        try {
+            refreshPool();
+        } catch (Throwable t) {
+            log.warn(
+                    "Failed to refresh Cassandra KVS pool."
+                            + " Extended periods of being unable to refresh will cause perf degradation.",
+                    t);
+        }
+
+        if (getCurrentPools().isEmpty()) {
+            log.error(
+                    "There are no pools remaining after refreshing and validating pools. Scheduling the next refresh"
+                            + " very soon to avoid an extended downtime.",
+                    SafeArg.of("attempt", attempt));
+            int maxShift = Math.min(MAX_ATTEMPTS_BEFORE_CAPPING_BACKOFF, attempt);
+            refreshPoolFuture = refreshDaemon.schedule(
```

Comment:
does `refreshPoolFuture` need to be `volatile` for safe publication and accesses?

Reply:
Yeah, have that and a few other changes to push up still :) Thanks for a look over, though!